zebra-chain/src/orchard/commitment.rs

```diff
@@ -0,0 +1,468 @@
+
+/// LEBS2OSP256(repr_P(cv))
+///
+/// https://zips.z.cash/protocol/protocol.pdf#pallasandvesta
+impl From<ValueCommitment> for [u8; 32] {
+    fn from(cm: ValueCommitment) -> [u8; 32] {
+        cm.0.to_bytes()
+    }
+}
+
+impl<'a> std::ops::Sub<&'a ValueCommitment> for ValueCommitment {
+    type Output = Self;
+
+    fn sub(self, rhs: &'a ValueCommitment) -> Self::Output {
+        self - *rhs
+    }
+}
+
+impl std::ops::Sub<ValueCommitment> for ValueCommitment {
+    type Output = Self;
+
+    fn sub(self, rhs: ValueCommitment) -> Self::Output {
+        ValueCommitment((self.0 - rhs.0).into())
+    }
+}
+
+impl std::ops::SubAssign<ValueCommitment> for ValueCommitment {
+    fn sub_assign(&mut self, rhs: ValueCommitment) {
+        *self = *self - rhs;
+    }
+}
+
+impl std::iter::Sum for ValueCommitment {
+    fn sum<I>(iter: I) -> Self
+    where
+        I: Iterator<Item = Self>,
+    {
+        iter.fold(
+            ValueCommitment(pallas::Affine::identity()),
+            std::ops::Add::add,
+        )
+    }
+}
+
+/// LEBS2OSP256(repr_P(cv))
+///
+/// https://zips.z.cash/protocol/protocol.pdf#pallasandvesta
+impl TryFrom<[u8; 32]> for ValueCommitment {
+    type Error = &'static str;
+
+    fn try_from(bytes: [u8; 32]) -> Result<Self, Self::Error> {
+        let possible_point = pallas::Affine::from_bytes(&bytes);
+
+        if possible_point.is_some().into() {
+            Ok(Self(possible_point.unwrap()))
+        } else {
+            Err("Invalid pallas::Affine value")
+        }
+    }
+}
+
+impl ZcashSerialize for ValueCommitment {
+    fn zcash_serialize<W: io::Write>(&self, mut writer: W) -> Result<(), io::Error> {
+        writer.write_all(&<[u8; 32]>::from(*self)[..])?;
+        Ok(())
+    }
+}
+
+impl ZcashDeserialize for ValueCommitment {
+    fn zcash_deserialize<R: io::Read>(mut reader: R) -> Result<Self, SerializationError> {
+        Self::try_from(reader.read_32_bytes()?).map_err(|e| SerializationError::Parse(e))
+    }
+}
+
+impl ValueCommitment {
+    /// Generate a new _ValueCommitment_.
+    ///
+    /// https://zips.z.cash/protocol/protocol.pdf#concretehomomorphiccommit
+    pub fn randomized<T>(csprng: &mut T, value: Amount) -> Self
+    where
+        T: RngCore + CryptoRng,
+    {
+        let rcv = generate_trapdoor(csprng);
+
+        Self::new(rcv, value)
+    }
+
+    /// Generate a new _ValueCommitment_ from an existing _rcv_ on a _value_.
+    ///
+    /// ValueCommit^Orchard(v) :=
+    ///
+    /// https://zips.z.cash/protocol/protocol.pdf#concretehomomorphiccommit
+    #[allow(non_snake_case)]
+    pub fn new(rcv: pallas::Scalar, value: Amount) -> Self {
+        let v = pallas::Scalar::from(value);
+
+        // TODO: These generator points can be generated once somewhere else to
+        // avoid having to recompute them on every new commitment.
+        let V = pallas_group_hash(b"z.cash:Orchard-cv", b"v");
+        let R = pallas_group_hash(b"z.cash:Orchard-cv", b"r");
```

Comment:
Let's just move them into a `lazy_static!` ?

Reply:
`lazy_static`ing these somehow makes the trait impls not observable?

Reply:
Lazy static creates a wrapper type that Derefs to the specified type. So you'll need to use the * operator, deref method, or trigger an implicit deref to get to the trait impl.

https://docs.rs/lazy_static/1.4.0/lazy_static/#semantics
https://doc.rust-lang.org/std/ops/trait.Deref.html

Reply:
Sometimes as_ref is also useful.

Here are some examples of complex lazy_static code in Zebra:
https://github.com/ZcashFoundation/zebra/blob/main/zebra-test/src/vectors/block.rs

Reply:
Thank you! I got this working, and in answering my question on whether it's wise to have a `lazy_static!{}` inside a function/method, turns out it's good because it limits scope for these statics to this scope alone 😎

<img width="707" alt="Screen Shot 2021-04-26 at 2 29 13 PM" src="https://user-images.githubusercontent.com/552961/116132648-c7f81c80-a69b-11eb-83d0-fd9762301ec6.png">https://users.rust-lang.org/t/lazy-static-inside-a-functions-scope/31700/2

Reply:
Resolved in b848cb37